### ISSUE TITLE
[Feature]: IntanRecordingInterface - digital channels and aux channels

### ISSUE THREAD

**Issue by jeongjunjjkim** [catalystneuro/neuroconv#519]
### What would you like to see added to NeuroConv?

Currently the IntanRecordingInterface only reads in the analog channels for electrodes, and is not reading in the digital channels, aux channels or adc channels, which often have useful synchronization signals (similar to NIDQ interface). Adding in a feature to read in the additional channels would be useful.

### Do you have any interest in helping implement the feature?

Yes, but I would need guidance.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/catalystneuro/neuroconv/blob/main/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this bug was not already [reported](https://github.com/catalystneuro/neuroconv/issues)?

**Comment by jeongjunjjkim** [comment]
spikeinterface.extractors.IntanRecordingExtractor contains analog channels, aux channels, supply voltage, and ADC channels, but completely misses digital channels. I've shared the google drive link to a sample rhd file via slack.

**Comment by CodyCBakerPhD** [comment]
@jeongjunjjkim the full `neo` header looks as follows on the shared file

```
{'nb_block': 1,
 'nb_segment': [1],
 'signal_streams': array([('RHD2000 amplifier channel', '0'),
        ('RHD2000 auxiliary input channel', '1'),
        ('RHD2000 supply voltage channel', '2'),
        ('USB board ADC input channel', '3')],
       dtype=[('name', '<U64'), ('id', '<U64')]),
 'signal_channels': array([('A-000', '0', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-001', '1', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-002', '2', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-003', '3', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-004', '4', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-005', '5', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-006', '6', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-007', '7', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-008', '8', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-009', '9', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-010', '10', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-011', '11', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-012', '12', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-013', '13', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-014', '14', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-015', '15', 20000.        , 'uint16', 'uV', 1.9500e-01, -6389.76, '0'),
        ('A-AUX1', '16',  5000.        , 'uint16', 'V', 3.7400e-05,     0.  , '1'),
        ('A-AUX2', '17',  5000.        , 'uint16', 'V', 3.7400e-05,     0.  , '1'),
        ('A-AUX3', '18',  5000.        , 'uint16', 'V', 3.7400e-05,     0.  , '1'),
        ('A-VDD1', '19',   333.33333333, 'uint16', 'V', 7.4800e-05,     0.  , '2'),
        ('ADC-00', '20', 20000.        , 'uint16', 'V', 5.0354e-05,     0.  , '3'),
        ('ADC-01', '21', 20000.        , 'uint16', 'V', 5.0354e-05,     0.  , '3'),
        ('ADC-02', '22', 20000.        , 'uint16', 'V', 5.0354e-05,     0.  , '3'),
        ('ADC-03', '23', 20000.        , 'uint16', 'V', 5.0354e-05,     0.  , '3'),
        ('ADC-04', '24', 20000.        , 'uint16', 'V', 5.0354e-05,     0.  , '3'),
        ('ADC-05', '25', 20000.        , 'uint16', 'V', 5.0354e-05,     0.  , '3')],
       dtype=[('name', '<U64'), ('id', '<U64'), ('sampling_rate', '<f8'), ('dtype', '<U16'), ('units', '<U64'), ('gain', '<f8'), ('offset', '<f8'), ('stream_id', '<U64')]),
 'spike_channels': array([],
       dtype=[('name', '<U64'), ('id', '<U64'), ('wf_units', '<U64'), ('wf_gain', '<f8'), ('wf_offset', '<f8'), ('wf_left_sweep', '<i8'), ('wf_sampling_rate', '<f8')]),
 'event_channels': array([], dtype=[('name', '<U64'), ('id', '<U64'), ('type', 'S5')])}
```

Just to 100% confirm - you expected there to be another channel there for digital input? What is the name that you expect that channel to have?

**Comment by jeongjunjjkim** [comment]
Yeah there should be a digital input channel, which is captured by the Intan Matlab reader. Should be called 'DIN-00'

**Comment by CodyCBakerPhD** [comment]
Great, thanksl This is an issue with SI or neo then so I'll raise it upstream

**Comment by CodyCBakerPhD** [comment]
@weiglszonja needs an `IntanDigitalInterface` for one of their current projects and so has been assigned to this issue now that the neo PR is merged: https://github.com/NeuralEnsemble/python-neo/pull/1249

**Comment by vigji** [comment]
I don't know if this is the right place to ask, but I'll try before opening a possibly duplicated issue.

I have a setup where an Intan board is used just to record digital signals related to time synching and stimulus logging that I would like to read into my nwb file. As far as I understand, currently the Intan data source is designed to be a source of ephys data, correct? 

I am currently using custom code to read the data arrays manually to just add them separately to the nwb file, but I would like to load them using existing pipelines.

I also have another issue with `.rhd` file reading that should be probably raised separately, but since I'm here: all my data are saved in multiple `.rhd` files, one minute long each, and I would need to concatenate them. The intan scripts that I'm using currently support this, and I see they have a concatenation option here https://github.com/Intan-Technologies/IntanToNWB but I am not sure I see the option for the current support to Intan data in `neuroconv`

**Comment by weiglszonja** [comment]
Hey @vigji, 

> As far as I understand, currently the Intan data source is designed to be a source of ephys data, correct?

You're indeed correct, although you can use [IntanRawIO](https://github.com/NeuralEnsemble/python-neo/blob/8ec58f7dd13a65f233cc89497e01efa572739bb2/neo/rawio/intanrawio.py#L34) to access the digital signals now.

> all my data are saved in multiple .rhd files, one minute long each, and I would need to concatenate them.

I've seen an example before where I needed to concatenate Intan recordings over time. I put together a custom interface [here](https://github.com/catalystneuro/turner-lab-to-nwb/blob/bdd197ed6d2adaafc44cae4f82c74c12c68bcd54/src/turner_lab_to_nwb/asap_embargo/asap_embargo_intan_recordinginterface.py), that uses `ConcatenateSegmentRecording` from `spikeinterface` to concatenate multiple `IntanRecordingExtractor` extractors.

Do you think @CodyCBakerPhD it would be worth having an `IntanConcatenatedRecordingInterface` added to neuroconv as well? This is the second example I've seen, but maybe you know more.

**Comment by CodyCBakerPhD** [comment]
@vigji Yep, this is the thread for this.

We do plan to eventually support auxiliary channels in general for anything the SpikeInterface extractors can read. Currently, this is only implemented for NIDQ in core NeuroConv

> Do you think @CodyCBakerPhD it would be worth having an IntanConcatenatedRecordingInterface added to neuroconv as well? This is the second example I've seen, but maybe you know more.

I leave that to @h-mayorquin to confirm, who has been pushing that upstream for one of his latest projects

**Comment by CodyCBakerPhD** [comment]
Related: https://github.com/catalystneuro/neuroconv/pull/520

**Comment by h-mayorquin** [comment]
@CodyCBakerPhD @weiglszonja 
The changes are for a different use case though. What I am adding is support for the "one-file-per-signal" and "one-file-per-channel" formats of Intan:

https://intantech.com/files/Intan_RHD2000_data_file_formats.pdf

This is different from concatenating across time as @weiglszonja is describing. I am missing some context here.

I think we are deviating too much from the original topic of this issue. We should discuss the context for this in another issue or in the next meeting among ourselves.

**Comment by CodyCBakerPhD** [comment]
> This is different from concatenating across time as @weiglszonja is describing. I am missing some context here.

Ah. So in all those different cases we have multiple Intan files meaning different things? depending on how axes should be aggregated? And no top-level header file (like we see in Bruker or other microscopy formats that follow this kind of pattern) telling us how to go about that to recover a frames x channel structure?

> I think we are deviating too much from the original topic of this issue. We should discuss the context for this in another issue or in the next meeting among ourselves.

Sure, this issue can be for aux channels - can you open another one for various concatenation types?

**Comment by h-mayorquin** [comment]
Once https://github.com/NeuralEnsemble/python-neo/pull/1491 is merged on neo and the digital channels are demultiplex, we can use the intan recording interface in spikeinterface to sort this one. @luiztauffer

**Comment by h-mayorquin** [comment]
Hi, brief status update on this.

Aux and ADC channels are now supported through `IntanAnalogInterface` (PR #1440), which also covers DC amplifier streams (RHS) and ADC outputs. RHS stimulation channels are handled by `IntanStimInterface` (PR #1711). PR #1724  adds support for sessions saved using Intan RHX's "create a new save file every N minutes" rotation option, where one session is split across several rotated `.rhd`/`.rhs` files in one folder.

The remaining piece is digital input/output channels. These are TTL signals that might semantically represent events rather than continuous traces, so the right NWB representation is the new built-in events data model proposed in [NWBEP001: Events](https://github.com/nwb-extensions/nwbep-review/issues/4). I am waiting on the release of that proposal in pynwb before implementing `IntanDigitalInterface`, because building against the current `ndx-events` extension and then rewriting against the new model would be duplicated work. Once NWBEP001 ships, the digital interface is the only remaining piece needed to fully close this issue and we will be in full intan support.

Meanwhile, I will be adding a converter to streamline intan data addition with a single class.

**Comment by h-mayorquin** [comment]
Ok, this in as a PR now. 
https://github.com/catalystneuro/neuroconv/pull/1812
